KomodoPlatform/komodo-defi-framework
# feat(trading-proto-upgrade): swap UTXO PoC using Storable State Machine.

#1895 

What's done:

- Implemented Storable State Machine abstraction with a goal to have as fewer changes to existing state machines as possible.
- Implemented successful swap of UTXO to UTXO coin. Tested using dockerized komodod daemons.
- Swap V2 message exchange is done using Protobuf. Fns like `broadcast_swap_v2_message` can (and likely should) be used for other protocols utilizing protobuf, but I preferred to keep them in `lp_swap` for now to not overlap with P2P upgrade PR.
- Did few other minor refactorings.

Important notes:

- It covers only success case now, no cancellation/refund is currently available.
- It uses dummy in-memory storage, swap data is not persisted for now.
- Swap status is not currently available through RPC API.
- Additional work is required in terms of swap negotiation steps, required validation of messages, etc.
- Seednodes should be updated to support/rebroadcast new swap protocol messages. 

Test plan:
1. Use only test coins or very small amounts of assets with real market value!
1. Run two nodes (seed and light) with `"use_trading_proto_v2": true` in MM2.conf.
1. Select coins to trade, both must have `"protocol": { "type": "UTXO" }`.
1. Initiate trading using `setprice` and `buy/sell` calls.
1. Following messages should be logged:
    ```
    05 04:28:54, mm2_main::mm2::lp_swap::maker_swap_v2:203] INFO Maker swap 36f2e374-9f79-415b-87ff- f1b1622cf7fd has successfully started
    ```
    ```
    05 04:28:54, mm2_main::mm2::lp_swap::taker_swap_v2:197] INFO Taker swap 36f2e374-9f79-415b-87ff-f1b1622cf7fd has successfully started
    ```
1. Swap transactions info is available only via logs now, e.g.
    ```
    05 04:28:57, mm2_main::mm2::lp_swap::maker_swap_v2:433] INFO Sent maker payment MYCOIN tx d5e3c9abd522eaf68e5ba699547632e27fd77985d1b3d56228b24ec7544a4dcd during swap 36f2e374-9f79-415b-87ff-f1b1622cf7fd
    ```
1. Swap should complete on both nodes, confirmed by log messages like
    ```
    05 04:49:39, mm2_main::mm2::lp_swap::taker_swap_v2:922] INFO Swap 3d4c27a6-1fc5-486c-a340-f9b3889f119a has been completed
    ```
